### PR TITLE
DRep expiry update with number of dormant epochs

### PR DESCRIPTION
**Pull request by aniketd**
# Description

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs [hunk 1]
```diff
@@ -254,3 +254,24 @@ instance
...
+  if numDormantEpochs == EpochNo 0
+    then drepState
+    else
+      drepState
+        <&> drepExpiryL
+          %~ ( \e ->
+                if e < currentEpoch
+                  then e
+                  else binOpEpochNo (+) numDormantEpochs e
+             )
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs]: We should not update the expiry only if DRep has already expired. But the current value in the `drepExpiry` does not reflect the actual value. The actual value is `drepExpiry +  numDormantEpochs`
```suggestion
          %~ ( \curExpiry ->
                let newExpiry = binOpEpochNo (+) numDormantEpochs curExpiry
                if newExpiry < currentEpoch
                  then curExpiry
                  else newExpiry
             )
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs [hunk 2]
```diff
@@ -207,7 +213,11 @@ conwayGovCertTransition = do
               Map.adjust
                 ( \drepState ->
                     drepState
-                      & drepExpiryL .~ addEpochInterval cgceCurrentEpoch ppDRepActivity
+                      & drepExpiryL
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs]: This scenario didn't even occur, it is very nice that you thought of this.

Can you also check that we do the same thing where we refresh the DRep expiry on a vote submission.

I can't seem to find where we do that, which might be a bug if it is missing.

But the idea is that it is possible scenario:
* there to be no proposals when the epoch starts and the number of dormant epochs is `numDormantEpochs > 0`
* someone submits a proposal
* a DRep votes on that proposal, thus refreshing its expiry `curEpoch + ppDRepActivity`
* epoch boundary comes we incorrectly update the expiry for that DRep to be further in the future than it has to be: `curEpoch + ppDRepActivity + numDormantEpochs`

In other words similar problem as with update certificate, except it needs more conditions to occur.

**Review by lehins — COMMENTED** [review]
Haven't reviewed it thoroughly yet, but here are a couple of important comments

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs (hunk 1 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs]: :bow: brilliant catch!

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs (hunk 2 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs]: Awesome catch again :bow:

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs [hunk 3]
```diff
@@ -204,14 +211,8 @@ conwayGovCertTransition = do
...
           { vsDReps =
-              Map.adjust
-                ( \drepState ->
-                    drepState
-                      & drepExpiryL .~ addEpochInterval cgceCurrentEpoch ppDRepActivity
-                      & drepAnchorL .~ mAnchor
-                )
-                cred
-                vsDReps
+              updateDRepExpiry ppDRepActivity cgceCurrentEpoch (vState ^. vsNumDormantEpochsL) cred $
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs]: Is it necessary to have a separate `Map.adjust` in `updateDRepExpiry` , rather than having it return the EpochNo  as `binOpEpochNo (-) (addEpochInterval currentEpoch ppDRepActivity) numDormantEpochs`  and then setting it here with `drepExpiryL` ?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs [hunk 4]
```diff
@@ -254,3 +254,28 @@ instance
...
+-- And also reset the `numDormantEpochs` counter.
+updateDormantDRepExpiry ::
+  -- | Current Epoch
+  EpochNo ->
+  -- | Number of dormant epochs
+  EpochNo ->
+  VState era ->
+  VState era
+updateDormantDRepExpiry currentEpoch numDormantEpochs vState =
+  if numDormantEpochs == EpochNo 0
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:265]: We are already checking this condition before calling the function, so it gets invoked only if numDormantEpochs /= EpochNo 0 , no?  Is there a point to check both before and within the function?

Comment (teodanciu) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:265]: If we're checking inside the function, then maybe we don't need to pass the `numDormantEpochs` as a parameter, since it's already inside the VState?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs [hunk 5]
```diff
@@ -254,3 +254,28 @@ instance
...
+    else
+      vState
+        & vsNumDormantEpochsL .~ EpochNo 0
+        & vsDRepsL %~ fmap updateExpiry
+  where
+    updateExpiry =
+      drepExpiryL
+        %~ \currentExpiry ->
+          let actualExpiry = binOpEpochNo (+) numDormantEpochs currentExpiry
+           in if actualExpiry < currentEpoch
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:277]: Just for my benefit, the difference from the previous implementation for the case in which we have proposals  and non-zero dormant epochs is here, right?  We don't extend the expiry if it's already expired, am I understanding correctly?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs [hunk 6]
```diff
@@ -227,3 +228,23 @@ conwayGovCertTransition = do
...
+  -- | DRep map from DRepState
+  Map.Map (Credential 'DRepRole c) (DRepState c) ->
+  Map.Map (Credential 'DRepRole c) (DRepState c)
+updateDRepExpiry ppDRepActivity currentEpoch numDormantEpochs =
+  Map.adjust $
+    drepExpiryL
+      .~ binOpEpochNo
+        (-)
+        (addEpochInterval currentEpoch ppDRepActivity)
+        numDormantEpochs
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs]: Won't this overflow for big `numDormantEpochs` ?

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs [hunk 7]
```diff
@@ -209,27 +210,26 @@ conwayCertsTransition = do
...
             let hasProposals = not . OSet.null $ tx ^. bodyTxL . proposalProceduresTxBodyL
                 numDormantEpochs = certState ^. certVStateL . vsNumDormantEpochsL
-                isNumDormantEpochsNonZero = numDormantEpochs /= EpochNo 0
-             in if hasProposals && isNumDormantEpochsNonZero
-                  then
-                    certState
-                      & certVStateL . vsDRepsL %~ (<&> (drepExpiryL %~ binOpEpochNo (+) numDormantEpochs))
-                      & certVStateL . vsNumDormantEpochsL .~ EpochNo 0
+             in if hasProposals && numDormantEpochs /= EpochNo 0
+                  then certState & certVStateL %~ updateDormantDRepExpiry currentEpoch numDormantEpochs
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs]: Slight simplification:
`updateDormantDRepExpiry` can lookup `vsNumDormantEpochsL`, we don't need to extract it and pass it as an argument.
```suggestion
             in if hasProposals
                  then certState & certVStateL %~ updateDormantDRepExpiry currentEpoch
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs [hunk 8]
```diff
@@ -209,27 +210,26 @@ conwayCertsTransition = do
...
                   else certState
 
       -- Update DRep expiry for all DReps that are voting in this transaction.
       -- This will execute in mutual-exclusion to the previous updates to DRep expiry,
       -- because if there are no proposals to vote on , there will be no votes either.
-      let drepActivity = pp ^. ppDRepActivityL
+      let numDormantEpochs1 = certState' ^. certVStateL . vsNumDormantEpochsL
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs]: Other binding with the same name will go away due to my suggestion.
```suggestion
      let numDormantEpochs = certState' ^. certVStateL . vsNumDormantEpochsL
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs [hunk 9]
```diff
@@ -209,27 +210,26 @@ conwayCertsTransition = do
...
+                  DRepVoter cred ->
+                    updateDRepExpiry drepActivity currentEpoch numDormantEpochs1 cred dreps
                   _ -> dreps
               )
               (certState' ^. certVStateL . vsDRepsL)
               (unVotingProcedures $ tx ^. bodyTxL . votingProceduresTxBodyL)
-          certStateWithDRepExpiryUpdated = certState' & certVStateL . vsDRepsL .~ updatedVSDReps
+
+      -- Final CertState with updates to DRep expiry based on new proposals and votes on existing proposals
+      let certStateWithDRepExpiryUpdated = certState' & certVStateL . vsDRepsL .~ updatedVSDReps
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs]: Let's change this to a modification lens. this way we can avoid duplicating this above:` (certState' ^. certVStateL . vsDRepsL)`
```suggestion
      let certStateWithDRepExpiryUpdated = certState' & certVStateL . vsDRepsL %~ updateVSDReps
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs (hunk 3 above)
Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs]: Yeah, I also agree with you. Moreover, I think it will be better to turn `updateDRepExpiry` to handle just one DRepState, rather than also do the `adjust` in a Map. This way it will look like that:
```haskell
 Map.adjust
   (\drepState -> 
     updateDRepExpiry
       ppDRepActivity
       cgceCurrentEpoch
       (vState ^. vsNumDormantEpochsL)
       (drepState & drepAnchorL .~ mAnchor))
   cred
   vsDReps

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs (hunk 6 above)
Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs]: It would indeed overflow for large enough `numDormantEpochs`, except `numDormantEpochs` is incremented with each epoch and is not a user supplied value, so it has no chance of overflowing.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs (hunk 4 above)
Comment (lehins) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:265]: Very nice suggestion, I noticed the same thing

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 10]
```diff
@@ -1377,22 +1415,47 @@ submitConstitution prevGovId = do
   govActionId <- submitGovAction constitutionAction
   pure (govActionId, constitution)
 
-expectExtraDRepExpiry ::
-  (HasCallStack, EraGov era, ConwayEraPParams era) =>
+expectDRepResigned ::
+  HasCallStack =>
+  Credential 'DRepRole (EraCrypto era) ->
+  ImpTestM era ()
+expectDRepResigned drep = do
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs]: DReps do not resign, it is CC members that can resign. We should avoid conflicting terminology.
```suggestion
expectDRepNotRegistered ::
  HasCallStack =>
  Credential 'DRepRole (EraCrypto era) ->
  ImpTestM era ()
expectDRepNotRegistered drep = do
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 11]
```diff
@@ -362,6 +371,35 @@ registerDRep = do
...
+  ( ShelleyEraImp era
+  , ConwayEraTxCert era
+  ) =>
+  Credential 'DRepRole (EraCrypto era) ->
+  ImpTestM era ()
+updateDRep drep =
+  submitTxAnn_ "Update DRep" $
+    mkBasicTx mkBasicTxBody
+      & bodyTxL . certsTxBodyL
+        .~ SSeq.singleton (UpdateDRepTxCert drep SNothing)
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs]: Let's add a bit of randomness in:
```suggestion
updateDRep drep = do
  mAnchor <- arbitrary
  submitTxAnn_ "Update DRep" $
    mkBasicTx mkBasicTxBody
      & bodyTxL . certsTxBodyL
        .~ SSeq.singleton (UpdateDRepTxCert drep mAnchor)
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs [hunk 12]
```diff
@@ -146,59 +146,223 @@ dRepSpec =
             submitParameterChange SNothing $ def & ppuMinFeeAL .~ SJust (Coin 3000)
       expectNumDormantEpochs 0
 
-      -- epoch 0
+      -- epoch 0: we submit a proposal, pulser is empty
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs]: Pulser isn't relevant for expiry
```suggestion
      -- epoch 0: we submit a proposal
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs [hunk 13]
```diff
@@ -146,59 +146,223 @@ dRepSpec =
...
+      expectNumDormantEpochs 0
+      expectDRepExpiry drep 2
+    it "expiry updates are correct for a mixture of cases" $ do
+      modifyPParams $ \pp ->
+        pp
+          & ppGovActionLifetimeL .~ EpochInterval 2
+          & ppDRepActivityL .~ EpochInterval 4
+      (drep1, _, _) <- setupSingleDRep 1_000_000 -- Receives an expiry update transaction certificate
+      (drep2, _, _) <- setupSingleDRep 1_000_000 -- Turns inactive due to natural expiry
+      (drep3, _, _) <- setupSingleDRep 1_000_000 -- Resigns and gets deleted
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs]:
```suggestion
      (drep3, _, _) <- setupSingleDRep 1_000_000 -- Unregisters and gets deleted
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs [hunk 14]
```diff
@@ -146,59 +146,223 @@ dRepSpec =
...
+      expectDRepExpiry drep2 6
+
+      passEpoch -- entering epoch 6
+      expectCurrentProposals
+      expectPulserProposals
+      expectNumDormantEpochs 0
+      expectDRepExpiry drep1 9
+      expectDRepExpiry drep2 6
+
+      passEpoch -- entering epoch 6
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs]: Count is off in the comment. 
```suggestion
      passEpoch -- entering epoch 7
```

Maybe we can `passNEpochsChecking`, where it would be like `passNEpochs` except it would also accept an expectation that would be checked for every epoch being crossed. This way we could write:
```haskell
      passNEpochsChecking 3 $ do -- passing epoch 6 through 7
        expectCurrentProposals
        expectPulserProposals
        expectNumDormantEpochs 0
        expectDRepExpiry drep1 9
        expectDRepExpiry drep2 6
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs [hunk 15]
```diff
@@ -208,9 +208,9 @@ returnProposalDeposits removedProposals oldUMap =
 
 -- | When there have been zero governance proposals to vote on in the previous epoch
 -- increase the dormant-epoch counter by one.
-updateNumDormantEpochs :: DRepPulsingState era -> VState era -> VState era
-updateNumDormantEpochs pulser vState =
-  if dormantEpoch pulser
+updateNumDormantEpochs :: Proposals era -> VState era -> VState era
+updateNumDormantEpochs ps vState =
+  if null $ ps ^. pPropsL
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs]: I think we need to remove the expired proposals, since they cannot be voted on.
```suggestion
  if null $ OMap.filter (\gas -> curEpochNo <= gasExpiresAfter gas) (ps ^. pPropsL)
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs [hunk 16]
```diff
@@ -146,59 +146,223 @@ dRepSpec =
...
+      expectNumDormantEpochs 0
+      expectDRepExpiry drep1 9
+      expectDRepExpiry drep2 6
+
+      passEpoch -- entering epoch 7
+      expectCurrentProposals
+      expectPulserProposals
+      expectNumDormantEpochs 0
+      expectDRepExpiry drep1 9
+      expectDRepExpiry drep2 6
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/EpochSpec.hs]: That is an interesting case. Technically the proposal that is still in the system is expired, so no DRep can vote on it, but because there proposals set is not empty dormant epoch is not being incremented. 
I think we might need to do something about this edge case. I made a suggestion in the EPOCH rule about this
```suggestion
      -- there is still one proposal in the state, but it is expired:
      expectCurrentProposals
      expectPulserProposals
      expectNumDormantEpochs 1
      expectDRepExpiry drep1 9
      expectDRepExpiry drep2 6
```

**Review by lehins — COMMENTED** [review]
Looks like there is still a small issue with dormant epochs counting, which I noticed thanks to tests in this PR. 
Other than that and a few minor suggestions this PR looks very nice

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs (hunk 5 above)
Comment (aniketd) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs:277]: Yes, you are correct. :+1:

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs [hunk 17]
```diff
@@ -254,3 +253,27 @@ instance
...
+  EpochNo ->
+  VState era ->
+  VState era
+updateDormantDRepExpiry currentEpoch vState =
+  if numDormantEpochs == EpochNo 0
+    then vState
+    else
+      vState
+        & vsNumDormantEpochsL .~ EpochNo 0
+        & vsDRepsL %~ fmap updateExpiry
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Certs.hs]: `Functor` for `Map` uses lazy implementation of `map`.
```suggestion
        & vsDRepsL %~ Map.map updateExpiry
```

File: libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs [hunk 18]
```diff
@@ -628,6 +629,10 @@ vsCommitteeStateL = lens vsCommitteeState (\vs u -> vs {vsCommitteeState = u})
 vsNumDormantEpochsL :: Lens' (VState era) EpochNo
 vsNumDormantEpochsL = lens vsNumDormantEpochs (\vs u -> vs {vsNumDormantEpochs = u})
 
+vsActualDRepExpiry :: Credential 'DRepRole (EraCrypto era) -> VState era -> Maybe EpochNo
+vsActualDRepExpiry cred vs =
+  fmap (binOpEpochNo (+) (vsNumDormantEpochs vs) . drepExpiry) $ Map.lookup cred $ vsDReps vs
```
Comment (lehins) [inline libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs]: Slight simplification
```suggestion
  binOpEpochNo (+) (vsNumDormantEpochs vs) . drepExpiry <$> Map.lookup cred (vsDReps vs)
```

**Review by lehins — APPROVED** [review]
:+1: Nice. Thank you.
